File: frontend/src/topology/diagnostics.ts
import type { TopologyDiagnostics, TopologyGraph, TopologyLink, TopologyNode } from "../types/topology";

const GENERIC_PORTS = new Set(["", "unknown", "uplink", "wired", "down", "none", "n/a", "null", "ethernet"]);

function canonicalPort(value: unknown): string {
  const text = String(value || "").trim();
  if (!text) return "";
  const low = text.toLowerCase();
  if (low.startsWith("port") && /^\d+$/.test(low.slice(4))) return String(Number(low.slice(4)));
  if (/^\d+$/.test(text)) return String(Number(text));
  return text;
}

function physicalPort(value: unknown): string {
  const pid = canonicalPort(value);
  return !pid || GENERIC_PORTS.has(pid.toLowerCase()) ? "" : pid;
}

function looksLikeMac(value: unknown): boolean {
  const compact = String(value || "").toLowerCase().replace(/[^0-9a-f]/g, "");
  return compact.length === 12;
}

function specificPorts(link: TopologyLink): Record<string, string> {
  const out: Record<string, string> = {};
  const src = physicalPort((link.source_port as { portId?: string } | undefined)?.portId);
  const tgt = physicalPort((link.target_port as { portId?: string } | undefined)?.portId);
  if (src) out[link.source] = src;
  if (tgt) out[link.target] = tgt;
  return out;
}

function portsConflict(a: Record<string, string>, b: Record<string, string>): boolean {
  for (const [id, port] of Object.entries(a)) {
    if (b[id] && b[id] !== port) return true;
  }
  return false;
}

function duplicatePhysicalEdges(links: TopologyLink[]): number {
  const groups = new Map<string, TopologyLink[]>();
  for (const link of links) {
    if (link.link_type !== "wired" && link.link_type !== "discovered_partial") continue;
    if (link.discovery_method === "physical_downstream") continue;
    const key = [link.source, link.target].sort().join("|");
    groups.set(key, [...(groups.get(key) || []), link]);
  }
  let extras = 0;
  for (const group of groups.values()) {
    const clusters: TopologyLink[][] = [];
    for (const link of group) {
      const ports = specificPorts(link);
      const hit = clusters.find((cluster) => {
        const cports: Record<string, string> = {};
        for (const item of cluster) Object.assign(cports, specificPorts(item));
        return !portsConflict(ports, cports);
      });
      if (hit) hit.push(link);
      else clusters.push([link]);
    }
    for (const cluster of clusters) {
      if (cluster.length > 1) extras += cluster.length - 1;
    }
  }
  return extras;
}

function unresolvedIdentity(nodes: TopologyNode[]): number {
  return nodes.filter((node) => {
    if (node.managed && node.type === "meraki" && !looksLikeMac(node.label)) return false;
    const label = String(node.hostname || node.label || "").trim();
    return looksLikeMac(label) || label.toLowerCase().startsWith("unknown downstream") || label.toLowerCase().startsWith("discovered neighbor");
  }).length;
}

function duplicateChassisCandidates(graph: TopologyGraph): number {
  const fromDebug = graph.topology_debug?.duplicate_chassis;
  if (Array.isArray(fromDebug)) return fromDebug.length;
  const servers = graph.nodes.filter((n) => !n.managed && (n.subtype === "server" || n.device_class === "server"));
  const byName = new Map<string, string[]>();
  for (const node of servers) {
    const key = String(node.label || "").trim().toLowerCase().replace(/[^a-z0-9]+/g, "");
    if (!key || looksLikeMac(node.label)) continue;
    byName.set(key, [...(byName.get(key) || []), node.id]);
  }
  let count = 0;
  for (const ids of byName.values()) {
    if (new Set(ids).size > 1) count += 1;
  }
  return count;
}

export function computeDiagnostics(graph: TopologyGraph): TopologyDiagnostics {
  const wired = graph.links.filter((l) => l.link_type === "wired" || l.link_type === "discovered_partial");
  const conf = { high: 0, medium: 0, low: 0, unknown: 0 };
  for (const link of wired) {
    const key = String(link.confidence || "unknown").toLowerCase();
    if (key === "high" || key === "medium" || key === "low") conf[key] += 1;
    else conf.unknown += 1;
  }
  const macs = new Map<string, string[]>();
  const serials = new Map<string, string[]>();
  for (const node of graph.nodes) {
    const mac = String(node.metadata?.mac || node.metadata?.macAddress || "")
      .trim()
      .toLowerCase();
    if (mac) macs.set(mac, [...(macs.get(mac) || []), node.id]);
    const serial = String(node.serial || "").trim().toLowerCase();
    if (serial) serials.set(serial, [...(serials.get(serial) || []), node.id]);
  }
  let duplicates = 0;
  for (const ids of [...macs.values(), ...serials.values()]) {
    if (new Set(ids).size > 1) duplicates += 1;
  }
  const linked = new Set(graph.links.flatMap((l: TopologyLink) => [l.source, l.target]));
  const orphans = graph.nodes.filter(
    (n) => !linked.has(n.id) && !n.metadata?.topologyRoot && n.subtype !== "firewall"
  ).length;
  const unresolved = Array.isArray(graph.topology_debug?.unresolved_nodes) ? graph.topology_debug.unresolved_nodes.length : 0;
  return {
    physical_edges: wired.length,
    wireless_edges: graph.links.filter((l) => l.link_type === "wireless").length,
    high_confidence: conf.high,
    medium_confidence: conf.medium,
    low_confidence: conf.low,
    unknown_confidence: conf.unknown,
    unresolved_nodes: unresolved,
    duplicate_identities: duplicates,
    duplicate_chassis_candidates: duplicateChassisCandidates(graph),
    duplicate_physical_edges: duplicatePhysicalEdges(graph.links),
    unresolved_identity_count: unresolvedIdentity(graph.nodes),
    orphans,
    total_nodes: graph.nodes.length,
    managed_nodes: graph.nodes.filter((n) => n.managed).length,
  };
}

File: backend/app/services/diagnostics.py
"""Topology confidence and identity diagnostics."""

from __future__ import annotations

from collections import defaultdict
from typing import Any

from app.models.schemas import TopologyGraph, TopologyLink
from app.services.identity import looks_like_mac, physical_port_id
from app.services.physical_topology import detect_duplicate_chassis


def _specific_ports(link: TopologyLink) -> dict[str, str]:
    ports: dict[str, str] = {}
    for node_id, blob in ((link.source, link.source_port or {}), (link.target, link.target_port or {})):
        pid = physical_port_id((blob or {}).get("portId"))
        if node_id and pid:
            ports[node_id] = pid
    return ports


def _ports_conflict(left: dict[str, str], right: dict[str, str]) -> bool:
    for node_id, port in left.items():
        other = right.get(node_id)
        if other and other != port:
            return True
    return False


def _duplicate_physical_edges(links: list[TopologyLink]) -> tuple[int, list[dict[str, Any]]]:
    groups: dict[tuple[str, str], list[TopologyLink]] = defaultdict(list)
    for link in links:
        if link.link_type not in {"wired", "discovered_partial"}:
            continue
        if str(link.discovery_method or "") == "physical_downstream":
            continue
        pair = tuple(sorted((link.source, link.target)))
        groups[(pair[0], pair[1])].append(link)
    extras = 0
    details: list[dict[str, Any]] = []
    for pair, group in groups.items():
        clusters: list[list[TopologyLink]] = []
        for link in group:
            ports = _specific_ports(link)
            placed = False
            for cluster in clusters:
                cports: dict[str, str] = {}
                for item in cluster:
                    cports.update(_specific_ports(item))
                if _ports_conflict(ports, cports):
                    continue
                cluster.append(link)
                placed = True
                break
            if not placed:
                clusters.append([link])
        for cluster in clusters:
            if len(cluster) > 1:
                extras += len(cluster) - 1
                details.append(
                    {
                        "pair": [pair[0], pair[1]],
                        "count": len(cluster),
                        "ids": [item.id for item in cluster],
                    }
                )
    return extras, details


def _unresolved_identity(graph: TopologyGraph) -> tuple[int, list[dict[str, str]]]:
    items: list[dict[str, str]] = []
    for node in graph.nodes:
        label = str(node.hostname or node.label or "").strip()
        if node.managed and node.type == "meraki" and not looks_like_mac(label):
            continue
        weak = looks_like_mac(label) or label.lower().startswith("unknown downstream") or label.lower().startswith(
            "discovered neighbor"
        )
        if weak:
            items.append({"id": node.id, "label": label})
    return len(items), items


def compute_diagnostics(graph: TopologyGraph) -> dict[str, Any]:
    wired = [link for link in graph.links if link.link_type in {"wired", "discovered_partial"}]
    conf_counts = {"high": 0, "medium": 0, "low": 0, "unknown": 0}
    for link in wired:
        key = str(link.confidence or "unknown").lower()
        if key not in conf_counts:
            key = "unknown"
        conf_counts[key] += 1

    macs: dict[str, list[str]] = defaultdict(list)
    serials: dict[str, list[str]] = defaultdict(list)
    for node in graph.nodes:
        mac = str((node.metadata or {}).get("mac") or (node.metadata or {}).get("macAddress") or "").strip().lower()
        if mac:
            macs[mac].append(node.id)
        serial = str(node.serial or "").strip().lower()
        if serial:
            serials[serial].append(node.id)
    duplicates = []
    for key, ids in {**{f"mac:{k}": v for k, v in macs.items()}, **{f"serial:{k}": v for k, v in serials.items()}}.items():
        unique = sorted(set(ids))
        if len(unique) > 1:
            duplicates.append({"key": key, "node_ids": unique})

    linked = {link.source for link in graph.links} | {link.target for link in graph.links}
    orphans = [
        {"id": node.id, "label": node.hostname or node.label}
        for node in graph.nodes
        if node.id not in linked and not (node.metadata or {}).get("topologyRoot") and node.subtype != "firewall"
    ]
    unresolved = list((graph.topology_debug or {}).get("unresolved_nodes") or [])
    node_map = {node.id: node for node in graph.nodes}
    chassis_candidates = detect_duplicate_chassis(node_map, list(graph.links))
    dup_edges, dup_edge_details = _duplicate_physical_edges(list(graph.links))
    unresolved_identity, unresolved_identity_ids = _unresolved_identity(graph)
    return {
        "physical_edges": len(wired),
        "wireless_edges": len([link for link in graph.links if link.link_type == "wireless"]),
        "high_confidence": conf_counts["high"],
        "medium_confidence": conf_counts["medium"],
        "low_confidence": conf_counts["low"],
        "unknown_confidence": conf_counts["unknown"],
        "unresolved_nodes": len(unresolved),
        "duplicate_identities": len(duplicates),
        "duplicate_chassis_candidates": len(chassis_candidates),
        "duplicate_physical_edges": dup_edges,
        "unresolved_identity_count": unresolved_identity,
        "orphans": len(orphans),
        "total_nodes": len(graph.nodes),
        "managed_nodes": len([n for n in graph.nodes if n.managed]),
        "duplicates": duplicates,
        "orphan_ids": [item["id"] for item in orphans],
        "duplicate_chassis": chassis_candidates,
        "duplicate_physical_edge_groups": dup_edge_details,
        "unresolved_identities": unresolved_identity_ids,
    }
